File: leetcode/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        pos = {v: p for p, v in enumerate(inorder)}

        def build(p, q, n):
            if 0 == n: return None
            if 1 == n: return TreeNode(preorder[p])

            v = preorder[p]
            k = pos[v]
            d = k - q

            par = TreeNode(v)
            par.left = build(p+1, q, d)
            par.right = build(p+d+1, q+d+1, n-d-1)

            return par

        return build(0, 0, len(preorder))

File: leetcode/105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
 public:
  TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder)
  {
    for (int i = 0; i < inorder.size(); ++i)
      pos_[inorder[i]] = i;
    return build(preorder, inorder, 0, 0, inorder.size());
  }

  TreeNode* build(const vector<int>& preorder,
                  const vector<int>& inorder,
                  int p, int q, int len)
  {
    if (0 == len) return nullptr;
    if (1 == len) return new TreeNode(preorder[p]);

    int val = preorder[p];
    int k = pos_[val], d = k - q;

    TreeNode* par = new TreeNode(val);
    par->left = build(preorder, inorder, p + 1, q, d);
    par->right = build(preorder, inorder, p + d + 1, q + d + 1,
                       len - d - 1);
    return par;
  }

 private:
  map<int, int> pos_;
};
